File: examples/src/stats/StatsCore.ts
// ============================================================================
// StatsCore.ts — 统计均值计算核心层（TypeScript 示例）
//
// 对应模板：templates/NewModule/{Name}Core.ts.template
// 设计原则（见 skills/typescript-SKILL.md）：
//   - strict null check：null/undefined 显式守卫
//   - 哨兵契约：无效输入返回 NaN，不抛异常
//   - 零外部依赖：纯函数实现，可独立单元测试
// ============================================================================

/** 计算结果：成功返回值，失败返回 NaN（哨兵契约） */
type ComputeResult = number;

/**
 * 计算算术均值。
 *
 * @param values - 数值数组；null/undefined 或空数组返回 NaN
 * @returns 均值；无效输入返回 NaN
 */
export function mean(values: number[] | null | undefined): ComputeResult {
  if (values === null || values === undefined) {
    return NaN;
  }
  if (values.length === 0) {
    return NaN;
  }
  const clean = values.filter((v) => typeof v === "number" && Number.isFinite(v));
  if (clean.length === 0) {
    return NaN;
  }
  const result = clean.reduce((sum, v) => sum + v, 0) / clean.length;
  if (!Number.isFinite(result)) {
    return NaN;
  }
  return result;
}

/**
 * 计算加权均值。
 */
export function weightedMean(
  values: number[] | null | undefined,
  weights: number[] | null | undefined,
): ComputeResult {
  if (values === null || values === undefined || weights === null || weights === undefined) {
    return NaN;
  }
  if (values.length === 0 || weights.length === 0 || values.length !== weights.length) {
    return NaN;
  }
  // 与 mean 对齐：逐对过滤非有限值/权重，避免 NaN/Inf 毒化整个结果。
  let totalWeight = 0;
  let weightedSum = 0;
  for (let i = 0; i < values.length; i++) {
    const v = values[i];
    const w = weights[i];
    if (!Number.isFinite(v) || !Number.isFinite(w)) continue;
    totalWeight += w;
    weightedSum += v * w;
  }
  if (totalWeight === 0) return NaN;
  const result = weightedSum / totalWeight;
  if (!Number.isFinite(result)) return NaN;
  return result;
}

File: examples/src/stats/StatsCore.go
// ============================================================================
// StatsCore.go — 统计均值计算核心层（Go 示例）
// 对应模板：templates/NewModule/{Name}Core.go.template
// 设计原则（见 skills/go-SKILL.md）：
//   - 哨兵值 NaN：0 是有效值，用 NaN 表示"无效"
//   - 零 panic：无效输入返回 NaN，不抛异常
// ============================================================================
package stats

import "math"

// Mean 计算算术均值。nil/空切片返回 NaN。
func Mean(values []float64) float64 {
	if values == nil || len(values) == 0 {
		return math.NaN()
	}
	sum, count := 0.0, 0
	for _, v := range values {
		if math.IsNaN(v) || math.IsInf(v, 0) {
			continue
		}
		sum += v
		count++
	}
	if count == 0 {
		return math.NaN()
	}
	result := sum / float64(count)
	if math.IsInf(result, 0) || math.IsNaN(result) {
		return math.NaN()
	}
	return result
}

// WeightedMean 计算加权均值。nil/空/长度不匹配/权重全零返回 NaN。
func WeightedMean(values, weights []float64) float64 {
	if values == nil || weights == nil ||
		len(values) == 0 || len(weights) == 0 ||
		len(values) != len(weights) {
		return math.NaN()
	}
	totalWeight := 0.0
	for _, w := range weights {
		totalWeight += w
	}
	if totalWeight == 0 {
		return math.NaN()
	}
	sum := 0.0
	for i, v := range values {
		sum += v * weights[i]
	}
	result := sum / totalWeight
	if math.IsInf(result, 0) || math.IsNaN(result) {
		return math.NaN()
	}
	return result
}
